### PR TITLE
WIP: Metal Backend

### PR DESCRIPTION
**Pull request by IsaacMarovitz**
**Description:**
This PR aims to implement a fully complete Metal backend primarily targeted at Apple Silicon devices.

A new shader backend has been written to translate into MSL, which is then compiled by Metal for the target device. Future explorations could seek to eliminate this extra step by directly generating LLVM bytecode for shaders, as is being explored by projects such as [DXMT](https://github.com/3Shain/dxmt).

This backend seeks to take advantage of the shared memory architecture of Apple Silicon devices, and, as a result, all buffers are host-imported, which allows for frequent opportunities for CPU-side inline updates rather than having to rely on GPU-side blits that can add additional render pass splits or "fences".

Much of the tracking logic for buffers is identical to Vulkan, allowing for similar smart caching of converted buffers, dependency management between command buffers, and (when it's implemented) buffer mirrors, which can even further reduce the number of fences and render pass splits for games that frequently update buffers between draws.

**TODO:**
- [x] Argument buffer-based binding model
- [x] D24S8 support
- [ ] Bindless support
- [ ] Buffer mirrors
- [ ] Queries & Counters

**Current Progress:**
The following section has been collated into smaller sections due to the size of each table.

<details>
    <summary>Perfect</summary>

| Game | Result |
| ----- | ------ |
| Sonic Mania | Perfect |
| Stardew Valley | Perfect |
| Cuphead | Perfect |
| Scott Pilgrim vs The World: The Game | Perfect |
| Hollow Knight | Perfect |
| Celeste | Perfect |
| Undertale | Perfect |
| Deltarune | Perfect |
| Balatro | Perfect |
| FEZ | Perfect |
| Sea of Stars | Perfect |
| Phoenix Wright: Ace Attorney Trilogy | Perfect |
| Apollo Justice: Ace Attorney Trilogy | Perfect |
| Mario Kart 8 Deluxe | Perfect |
| Super Mario 3D World + Bowser's Fury | Perfect |
| Super Mario Bros. Wonder | Perfect |
| Super Mario Odyssey | Perfect |
| Super Mario Maker 2 | Perfect |
| Super Mario Galaxy | Perfect |
| Super Smash Bros. Ultimate | Perfect |
| Persona 5 Royal | Perfect |
| Luigi's Mansion 2 HD | Perfect |
| SEGA AGES Virtua Racing | Perfect |
| Sayonara Wild Hearts | Perfect |
| FAST RMX | Perfect |
| DARK SOULS: REMASTERED | Perfect |
| MONSTER HUNTER RISE | Perfect |
| The Legend of Zelda: Skyward Sword HD | Perfect |
| The Legend of Zelda: Link's Awakening | Perfect |
| Animal Crossing: New Horizons | Perfect (Aside from hardware bugs present on Vulkan) |
| Mario Party Superstars | Perfect (Aside from bindless effects) |
| WarioWare: Get It Together! | Much of the game does not render | 
</details>

<details>
    <summary>Partially Working</summary>

| Game | Result |
| ----- | ------ |
| DOOM (2016) | Small glitchy overlay on everything |
| The Legend of Zelda: Tears of the Kingdom | Problems with pop-in |
| The Legend of Zelda: Breath of the Wild | Sky flickering, game very slow |
| The Legend of Zelda: Echoes of Wisdom | Void does not render correctly |
| NieR:Automata The End of YoRHa Edition | Minor graphical problems |
| Hatsune Miku: Project DIVA Mega Mix | Intro video freezes |
| Kirby and the Forgotten Land | Some pop-in but otherwise perfect |
| Sonic Colors: Ultimate | Broken vertex shaders |
| Sonic Frontiers | Some lighting issues |
| Mario Strikers: Battle League | Crashes attempting layered rendering |
| Pokémon Legends: Arceus | Boots, freezes when loading into game |
| POKÉMON TOURNAMENT DX | Broken geometry |
| Xenoblade Chronicles: DE | Boots, freezes when loading into game |
| Super Mario RPG | Lighting problems |
| Super Mario 64 | Incorrect geometry | 
| Super Mario Sunshine | Game not displaying properly |
| Paper Mario: The Thousand-Year Door | Many rendering problems related to bindless |
| Penny's Big Breakaway | Many rendering problems due to lack of bindless |
| Metroid Prime Remastered | UI broken, game itself black, crashes due to layered rendering |
| Bayonetta 3 | Does not render correctly, broke similarly to OpenGL |
| Sphinx and the Cursed Mummy | Does not blit final framebuffer correctly |
| Fire Emblem: Three Houses | Some UI effect missing, GPU faults in game |
| LOLLIPOP CHAINSAW RePOP | Freezes in-game |
| Luigi's Mansion 3 | Crashes due to missing tesselation support |
</details>

<details>
    <summary>Renders Nothing</summary>

| Game | Result |
| ----- | ------ |
| Princess Peach: Showtime! | Crashes due to layered rendering |
</details>

Celeste

### CODE REVIEW

**Comment by amayra** [comment]
I remember reading about there many missing parts in metal for any emulators to work properly even though many modern game engines don't work because of that  

did apple improve metal API recently for Ryujinx to work natively in metal API

**Comment by IsaacMarovitz** [comment]
> did apple improve metal API recently for Ryujinx to work natively in metal API ?

Ryujinx already runs on Metal through MoltenVK and yes there is stuff we need to work around most notably Transform Feedback and Geometry Shaders. We already have workaround for these using MVK.

**Comment by ilyassemokhtari** [comment]
does it already work well ?

**Comment by IsaacMarovitz** [comment]
> does it already work well ?

If you glance at the checkboxes in the PR description, you will quickly see that this is a long way off from playing anything.

**Comment by ilyassemokhtari** [comment]
> > does it already work well ?
> 
> If you glance at the checkboxes in the PR description, you will quickly see that this is a long way off from playing anything.

Ahh. I see. Hope you can find working solutions for all of that. and thanks for sharing this with everybody

**Comment by duckfromdiscord** [comment]
this would be freaking amazing if it happened, no rush I know that this is an extremely difficult undertaking & i definitely do not expect to have a working game in 3 seconds, but id love to see where this goes. good luck 😄

**Comment by phoenixpereira** [comment]
@IsaacMarovitz Would any of the features announced today during WWDC2023 be helpful for this PR? These include a game porting toolkit, metal shader converter, offline shader compilation and frame pacing. I think the videos and sample code for these features get released later today if you want to check it out.

**Comment by IsaacMarovitz** [comment]
> @IsaacMarovitz Would any of the features announced today during WWDC2023 be helpful for this PR? These include a game porting toolkit, metal shader converter, offline shader compilation and frame pacing. I think the videos and sample code for these features get released later today if you want to check it out.

Those features are unfortunately not helpful for Ryujinx as they're designed around DirectX not Vulkan

**Comment by CoolFrenchPerson** [comment]
Would any of the changes made in 1.1.896 affect this pr ?

**Comment by IsaacMarovitz** [comment]
Letting y'all know this isn't dead yet. Did a lot of work on SharpMetal today and I think I can have something rendering with it soon, at which point I can work seriously on this PR.

**Comment by IsaacMarovitz** [comment]
SharpMetal is not quite ready for prime time but it’s pretty much 95% there, so I’ve started more serious work on this PR. The stuff left to implement is immense and we’re a ways off anything meaningful rendering, but I’m going to give it my best shot :)

**Comment by IsaacMarovitz** [comment]
image

First "present". Technically doesn't count cause the colour is completely wrong lol. It might be a while until even this basic sample is working. Stay tuned.

**Comment by IsaacMarovitz** [comment]
image

Texture `SetData` is working!

**Comment by IsaacMarovitz** [comment]
<h1>WARNING: EXTREME FLASHING/STROBE</h1>
<h2>Do **NOT** watch if you have photosensitive epilepsy or sensitive vision</h2>

https://github.com/Ryujinx/Ryujinx/assets/42140194/b17e5970-0a7e-4910-b60c-45b05c7235c2

First fully working prototype 🥳, expect Barry in his RGB glory soon. 🤞

**Comment by IsaacMarovitz** [comment]
https://github.com/Ryujinx/Ryujinx/assets/42140194/6dc45838-ae83-42bb-9b2d-e7347e6dcc8d

he has arrived

**Comment by IsaacMarovitz** [comment]
https://github.com/Ryujinx/Ryujinx/assets/42140194/b710ec49-d064-4cdf-860f-f18de724676b

he has arrived

**Comment by alessandro893** [comment]
Thank you very much for your  development.

**Comment by IsaacMarovitz** [comment]
Screenshot 2023-08-04 at 18 50 52

Starting to see some progress on shader gen 🤞

**Comment by jduncanator** [comment]
In regards to shader-gen. Metal binary shaders are just LLVM-IR. Might make better sense to target that directly instead of MSL first, which would significantly help reduce shader stutter and compilation time on Mac targets.

**Comment by IsaacMarovitz** [comment]
> In regards to shader-gen. Metal binary shaders are just LLVM-IR. Might make better sense to target that directly instead of MSL first, which would significantly help reduce shader stutter and compilation time on Mac targets.

Metal libraries are not documented by Apple, I am cautious taking this route as there's no guarantee this will work as intended. This is interesting and worth looking into tho.

**Comment by IsaacMarovitz** [comment]
image

Getting fairly close to working shaders 🤞

**Comment by IsaacMarovitz** [comment]
image

Simple valid shaders generating!

**Comment by IsaacMarovitz** [comment]
Shader gen (not complete) but stable

**Comment by IsaacMarovitz** [comment]
First commercial titles rendering

image

Screenshot 2024-03-20 at 08 01 26

**Comment by EricKY26** [comment]
> Letting y'all know this isn't dead yet. Did a lot of work on SharpMetal today and I think I can have something rendering with it soon, at which point I can work seriously on this PR.

is there any update lately? saw you merged lots of code, just out of curiosity

**Comment by Mika412** [comment]
@EricKY26 For this kind of questions please use Discord. If there's anything to share the OP will post it

**Comment by IsaacMarovitz** [comment]
> is there any update lately? saw you merged lots of code, just out of curiosity

The merged changes are upstream commits that included several changes to GAL, such as the large sampler arrays and bindless on Vulkan. I merged these in to reduce the number of conflicts so that this branch remains buildable.

I do not have any progress to share on the PR itself. I expect I will have time to work on it in a month or so as that is when I finish school.

My plan is to, in a substantial way, rewrite much of my existing work here with the exception of the shader backend. I plan to redo the binding model entirely to utilise argument buffers, which I was unsuccessful in getting to work in my initial local testing several weeks ago, and so was never committed.

As always, the limiting factor remains the lack of decent bindings. SharpMetal is *mostly* good enough, but it is missing some portions of the API that make it difficult to work with. I would either need to finish my [rewrite of SharpMetal](https://github.com/IsaacMarovitz/SharpMetal/pull/5), or wait for [Silk.NET to build their own bindings](https://github.com/dotnet/Silk.NET/issues/1121), which should happen at some point ™️ during the development of Silk 3.0.

**Comment by SamoZ256** [comment]
> > is there any update lately? saw you merged lots of code, just out of curiosity
> 
> The merged changes are upstream commits that included several changes to GAL, such as the large sampler arrays and bindless on Vulkan. I merged these in to reduce the number of conflicts so that this branch remains buildable.
> 
> I do not have any progress to share on the PR itself. I expect I will have time to work on it in a month or so as that is when I finish school.
> 
> My plan is to, in a substantial way, rewrite much of my existing work here with the exception of the shader backend. I plan to redo the binding model entirely to utilise argument buffers, which I was unsuccessful in getting to work in my initial local testing several weeks ago, and so was never committed.
> 
> As always, the limiting factor remains the lack of decent bindings. SharpMetal is _mostly_ good enough, but it is missing some portions of the API that make it difficult to work with. I would either need to finish my [rewrite of SharpMetal](https://github.com/IsaacMarovitz/SharpMetal/pull/5), or wait for [Silk.NET to build their own bindings](https://github.com/dotnet/Silk.NET/issues/1121), which should happen at some point ™️ during the development of Silk 3.0.

What’s the point of using argument buffers? Is it just to work around the limit of maximum bound buffers, textures and samplers in Metal?

**Comment by IsaacMarovitz** [comment]
> What’s the point of using argument buffers? Is it just to work around the limit of maximum bound buffers, textures and samplers in Metal?

TLDR; Yes, and it works out better with the way Ryujinx binds stuff to shaders.

**Comment by IsaacMarovitz** [comment]
Thanks to @SamoZ256, we have made significant progress!

Screenshot 2024-05-16 at 16 20 17

**Comment by IsaacMarovitz** [comment]
Odyssey Debug:

image

image

Font atlas is correct, but missing vertex buffer binding (storage buffer), so all draws are skipped.

image

Darkness issue not present on source texture issue with sRGB -> non sRGB.

Final framebuffer texture not correctly presented either.

**Comment by IsaacMarovitz** [comment]
image

Proud to announce that Mania is perfect!

**Comment by IsaacMarovitz** [comment]
I've resolved the formatting issues, so the commented build has been updated for the first time since May. It may work with many of your games if you want to take it for a test run. However, plenty is still known to be broken, so at this stage, I'm not looking for any genuine testing or comparisons against Vulkan.

Regardless, have fun, and I don't mind hearing about successes 😆

**Comment by stonerl** [comment]
@IsaacMarovitz I just noticed it got sometimes stuck when compiling the shaders. They were originally created using the Vulkan backend. I had to quit the app a couple of times, but with each iteration, it was able to compile more shaders. Now that all shaders have been compiled with the Metal backend. MK8D runs perfect.

Thanks for your work. BTW, Could we help with testing?

**Comment by epic-bruh-moment** [comment]
Tested Phoenix Wright Ace Attorney Trilogy. Game boots but shows a black screen with sound.

**Comment by IsaacMarovitz** [comment]
> Tested Phoenix Wright Ace Attorney Trilogy. Game boots but shows a black screen with sound.

Please try again on the latest commits.

**Comment by epic-bruh-moment** [comment]
> > Tested Phoenix Wright Ace Attorney Trilogy. Game boots but shows a black screen with sound.
> 
> Please try again on the latest commits.

Ace Attorney Trilogy now works perfectly.

I also tried Pokken tournament DX, initial main menu has severe lag and character models are not properly rendered
image

Super Mario RPG also works however many effects are either not rendered correctly or are not rendered outright
image
image

**Comment by tgcordell** [comment]
Really excited about all of the progress here! Was going to test out performance between this latest build and main branch but realized I dont have my controller handy.... Ended up having a crash after pausing and then stopping emulation on TOTK w/ the 1caa12e build on an M1 Pro. See attached. Confirmed the pop-in issues were still occurring.

[Ryujinx_1.1.0+1caa12e_2024-08-12_20-37-21.log](https://github.com/user-attachments/files/16594190/Ryujinx_1.1.0%2B1caa12e_2024-08-12_20-37-21.log)

**Comment by epic-bruh-moment** [comment]
Apollo Justice: Ace Attorney Trilogy is partially working. Certain effects such as the aluminium powder do not display as intended

image

**Comment by epic-bruh-moment** [comment]
Apollo Justice: Ace Attorney Trilogy is partially working, while most of the game runs perfectly, certain effects don't display and this one halts your progress. The aluminium dust does not appear properly on Metal

Behavior on Metal:
image

Expected Behavior on vulkan:
image

**Comment by epic-bruh-moment** [comment]
image
image

Issues still persist in Dual Destinies. Both of the Great Ace Attorney games however work perfectly.

**Comment by IsaacMarovitz** [comment]
@epic-bruh-moment Feel free to test titles that were not working again on latest commits, I expect them to resolve quite a few problems

**Comment by epic-bruh-moment** [comment]
> @epic-bruh-moment Feel free to test titles that were not working again on latest commits, I expect them to resolve quite a few problems

Apollo Justice now works perfectly on metal, however the issue with dual destinies still persists, I will test some other games later too.

**Comment by IsaacMarovitz** [comment]
@epic-bruh-moment Latest commit fixes Dual Destinies

image

**Comment by epic-bruh-moment** [comment]
> @epic-bruh-moment Latest commit fixes Dual Destinies
> 
> image

awesome, thanks a ton.

**Comment by epic-bruh-moment** [comment]
Pokken Tournament DX: Slightly better than before, while models are still messy, a lot more is now rendered and vfx is also working quite well
image

Super Mario: RPG: Certain areas are still darker than they should be however shadows are now rendering properly and are not flickering like crazy as they once were
image
image

The Great Ace Attorney Chronicles: Works Perfectly.
image

**Comment by iMonZ** [comment]
Wow that looks amazing. So exited trying this out.
